R/server.R

```diff
@@ -802,21 +830,41 @@ runExample <- function(example=NA,
 #'
 #' @export
 runGadget <- function(app, server = NULL, port = getOption("shiny.port"),
-  viewer = paneViewer()) {
+  viewer = paneViewer(), stopOnCancel = TRUE) {
 
   if (!is.shiny.appobj(app)) {
     app <- shinyApp(app, server)
   }
 
+  if (isTRUE(handleCancel)) {
+    app <- decorateServerFunc(app, function(input, output, session) {
+      observeEvent(input$cancel, {
+        stopApp(stop("User cancel"))
+      })
+    })
+  }
+
   if (is.null(viewer)) {
     viewer <- browseURL
   }
 
-  retVal <- withVisible(shiny::runApp(app, port = port, launch.browser = viewer))
-  if (retVal$visible)
-    retVal$value
-  else
-    invisible(retVal$value)
+  shiny::runApp(app, port = port, launch.browser = viewer)
+}
+
+# Add custom functionality to a Shiny app object's server func
+decorateServerFunc <- function(appobj, serverFunc) {
+  origServerFuncSource <- appobj$serverFuncSource
+  appobj$serverFuncSource <- function() {
+    origServerFunc <- origServerFuncSource()
+    function(input, output, session) {
+      serverFunc(input, output, session)
+      if (length(formals(plot)) == 2)
```

Comment:
I meant that `origServerFunc` needs to be checked not only that it has two arguments, but that they're named `input` and `output`.

This is the code I was referring to:
  https://github.com/rstudio/shiny/blob/af0463e/R/server.R#L305-L315

Although, I suppose that simply invoking with named arguments would be sufficient, as in:

```R
origServerFunc(input = input, output = output)

origServerFunc(input = input, output = output, server = server)
```

I think `clientData` probably isn't necessary since we decided that `session$clientData` is preferred.